### ISSUE TITLE
Running some scripts without any input sometimes leads users to believe that the program is hanging

### ISSUE THREAD

**Issue by AlexDaniel** [perl6/problem-solving#60]
Example:
* https://www.reddit.com/r/perl6/comments/bc0efq/sql2png_turn_create_table_statements_into_a_graph/ekq6khy/

Basically, any simple script that relies on `lines`, `words`, `slurp` and other subs can cause this. Like:
```perl6
sub MAIN() { .sqrt.say for lines }
```

When you run it without piping any data into it, it starts waiting for user input, which is not indicated in any way. While it is possible to say that it's a user error, or that the developer of that particular script could've added a message (`note ‘Type something here:’ if $*IN.t`), it is still something we can solve for everyone in the language/compiler.

**Comment by AlexDaniel** [comment]
One solution was proposed with the following commits:
* https://github.com/rakudo/rakudo/commit/ad8b5a649755abfe8290a5d77f4f53ecfc91df49
* https://github.com/rakudo/rakudo/commit/3060d1d0bd7e902d8ee501667df456b99dec2748
* https://github.com/rakudo/rakudo/commit/7668301dd60bffb1e3e6d180ecbfc1c4a476c260

It doesn't cover all cases (e.g. `sub MAIN() { loop { say get } }`), but I'm not sure if it should. Moreover, I'm interested in seeing how this can be reflected in the docs. These commits were reverted for the release to stay safe. That being said, I don't have very realistic examples on how this behavior can hurt. I tried various things (like ncurses), and there are cases when extra printing to stderr is unwanted, but the examples get very unrealistic.

**Comment by lizmat** [comment]
To explain a bit in text: the message would be shown if the following conditions were met:

1. a `MAIN` subroutine should be present
2. `$*IN`, `*$OUT`, `$*ERR` all have their `.t` to be `True`
3. no output on `$*OUT` or `$*ERR` has been made
4. subroutines `lines`, `words` or `slurp` are called

**Comment by jnthn** [comment]
No small part of understanding an imperative program is understanding the *effects* (and coeffects) of the things we invoke. The idea that something that we expect to read input could, as a side-effect, directly produce *output*, is highly undesirable from an API design and program reasoning perspective.

If something like the proposed solution goes in, the it should `warn` instead of `note`, which at least allows the language user the chance to stick in a `quietly` to suppress it. That said, I'm still not entirely comfortable with the set of heuristics at work here. For example, I think you can create the very same situation even without having a `MAIN`?

**Comment by lizmat** [comment]
Yes, you can create this situation without having a MAIN.  But then to have subs such as `lines` warn (or note), felt wrong.  If someone adds a MAIN, it at least feels like they want to have something more permanent / flexible than just a one-off script.  Also, when you use a MAIN, then the scope of using `lines` and friends is very clear.

But I guess I'm the only one who feels like this is worthwhile to have.  So by all means, close this issue if you feel it's not worthwhile.  I still stand by the fact that I think it's LTA for a script to silently wait for input on STDIN when there is a terminal attached.

**Comment by AlexDaniel** [comment]
@lizmat it's an interesting problem, and we should discuss what can be done about it.
